model_api/cpp/tilers/src/detection.cpp

```diff
@@ -0,0 +1,208 @@
+/*
+// Copyright (C) 2023 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+*/
+
+#include <algorithm>
+#include <vector>
+#include <opencv2/core.hpp>
+
+#include <tilers/detection.h>
+#include <models/results.h>
+#include <utils/nms.hpp>
+
+
+namespace {
+
+cv::Mat non_linear_normalization(cv::Mat& class_map) {
+    double min_soft_score, max_soft_score;
+    cv::minMaxLoc(class_map, &min_soft_score);
+    cv::pow(class_map - min_soft_score, 1.5, class_map);
+
+    cv::minMaxLoc(class_map, &min_soft_score, &max_soft_score);
+    class_map = 255.0 / (max_soft_score + 1e-12) * class_map;
+
+    return class_map;
+}
+
+}
+
+DetectionTiler::DetectionTiler(std::unique_ptr<ModelBase> _model, const ov::AnyMap& configuration) :
+    TilerBase(std::move(_model), configuration) {
+
+    auto ov_model = model->getModel();
+
+    auto max_pred_iter = configuration.find("max_pred_number");
+    if (max_pred_iter == configuration.end()) {
+        if (ov_model->has_rt_info("model_info", "max_pred_number")) {
+            max_pred_number = ov_model->get_rt_info<size_t>("model_info", "max_pred_number");
+        }
+    } else {
+        max_pred_number = max_pred_iter->second.as<size_t>();
+    }
+}
+
+std::unique_ptr<ResultBase> DetectionTiler::postprocess_tile(std::unique_ptr<ResultBase> tile_result, const cv::Rect& coord) {
+    DetectionResult* det_res = static_cast<DetectionResult*>(tile_result.get());
+    for (auto& det : det_res->objects) {
+        det.x += coord.x;
+        det.y += coord.y;
+    }
+
+    auto tmp_feature_vector = ov::Tensor(det_res->feature_vector.get_element_type(), det_res->feature_vector.get_shape());
+    auto tmp_saliency_map = ov::Tensor(det_res->saliency_map.get_element_type(), det_res->saliency_map.get_shape());
+
+    det_res->feature_vector.copy_to(tmp_feature_vector);
+    det_res->feature_vector = tmp_feature_vector;
+
+    det_res->saliency_map.copy_to(tmp_saliency_map);
+    det_res->saliency_map = tmp_saliency_map;
```

Comment:
Here's a suggestion. It's up to you to follow it.

Users of `DetectionTiler` will probably end up calling `wrap_saliency_map_tensor_to_mat()` similar to `merge_saliency_maps()`. In addition this will make them to consciously track the life time of underlying tensor. The proposal is to define a struct similar to `DetectionResult` but with `cv::Mat` as fields which would be constructed here in `postprocess_tile()` and already contain copied memory.

It looks that the only type mismatch which will be introduced by this solution is in tests, because the implementation is currently shared between model and tiler

Reply:
The same problem is valid for all the wrappers that output feature vectors and saliency maps.
I aimed to preserve the output type of the underlying detection model, since MediaPipe calculators already implemented and target the existing data types. Or do you mean this extra result structure can be implemented only for internal data transfer in Tiler?

Reply:
I don't think there's a big difference for internal data structure. Moreover you've already implemented it the other way